glean-core/android/src/test/java/mozilla/telemetry/glean/pings/CustomPingTest.kt

```diff
@@ -0,0 +1,109 @@
+package mozilla.telemetry.glean.scheduler
+
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import mozilla.telemetry.glean.Glean
+import mozilla.telemetry.glean.getContextWithMockedInfo
+import mozilla.telemetry.glean.getMockWebServer
+import mozilla.telemetry.glean.private.NoReasonCodes
+import mozilla.telemetry.glean.private.PingType
+import mozilla.telemetry.glean.resetGlean
+import mozilla.telemetry.glean.testing.GleanTestRule
+import mozilla.telemetry.glean.triggerWorkManager
+import mozilla.telemetry.glean.utils.tryGetLong
+import org.junit.Assert.assertEquals
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.json.JSONObject
+import java.util.concurrent.TimeUnit
+
+/**
+ * Testing behavior of custom pings.
+ *
+ * We already rely on the Rust side to test custom pings,
+ * but this enables us to test the upload mechanism specifically.
+ *
+ * Even if this seemingly duplicates some of the testing, this should be kept around.
+ */
+@RunWith(AndroidJUnit4::class)
+class CustomPingTest {
+    @get:Rule
+    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())
+
+    @Test
+    fun `sends empty custom ping`() {
+        // a smoke test for custom pings
+
+        val server = getMockWebServer()
+        val context = getContextWithMockedInfo()
+
+        resetGlean(context, Glean.configuration.copy(
+            serverEndpoint = "http://" + server.hostName + ":" + server.port,
+            logPings = true
+        ), clearStores = true, uploadEnabled = true)
+
+        // Define a new custom ping inline.
+        val customPing = PingType<NoReasonCodes>(
+            name = "custom-ping",
+            includeClientId = true,
+            sendIfEmpty = true,
+            reasonCodes = emptyList()
+        )
+
+        customPing.submit()
+        triggerWorkManager(context)
+
+        val request = server.takeRequest(2L, TimeUnit.SECONDS)
+        val docType = request.path.split("/")[3]
+        assertEquals("custom-ping", docType)
+    }
+
+    @Test
+    fun `multiple pings in one go`() {
+        val server = getMockWebServer()
+        val context = getContextWithMockedInfo()
+
+        resetGlean(context, Glean.configuration.copy(
+            serverEndpoint = "http://" + server.hostName + ":" + server.port,
+            logPings = true
+        ), clearStores = true, uploadEnabled = true)
+
+        // Define a new custom ping inline.
+        val customPing = PingType<NoReasonCodes>(
+            name = "custom-ping",
+            includeClientId = true,
+            sendIfEmpty = true,
+            reasonCodes = emptyList()
+        )
+
+        // Trigger the ping twice.
+        customPing.submit()
```

Comment:
Does it make sense to turn "test mode" off in these tests, to have async behaviour?

Reply:
Actually ... it might. On the other hand it might lead to other problems. I'll run it a few times locally.